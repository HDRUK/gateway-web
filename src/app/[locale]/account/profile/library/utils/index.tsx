import { IconButton } from "@mui/material";
import { createColumnHelper } from "@tanstack/react-table";
import { LibraryListItem, SelectedLibrary } from "@/interfaces/Library";
import Link from "@/components/Link";
import StyledCheckbox from "@/components/StyledCheckbox";
import TooltipIcon from "@/components/TooltipIcon";
import { CheckIcon, DeleteForeverIcon } from "@/consts/icons";
import { RouteName } from "@/consts/routeName";

const columnHelper = createColumnHelper<LibraryListItem>();

const getColumns = ({
    handleSelect,
    handleRemove,
    selected,
    translations,
}: {
    handleSelect: (data: SelectedLibrary) => void;
    handleRemove: (id: string | number) => void;
    selected: SelectedLibrary;
    translations: { [id: string]: string };
}) => [
    columnHelper.display({
        id: "actions",
        meta: { isPinned: true },
        cell: ({ row }) => {
            const {
                id,
                datasetId,
                name,
                dataCustodianId,
                dataCustodian,
                darEnabled,
                darTemplatePublished,
                cohortEnabled,
            } = row.original;
            return (
                <div style={{ textAlign: "center" }}>
                    <StyledCheckbox
                        checked={selected[id]?.selected}
                        onChange={(_e, value) =>
                            handleSelect({
                                [id]: {
                                    selected: value,
                                    datasetId,
                                    name,
                                    teamId: dataCustodianId,
                                    teamName: dataCustodian,
                                    darEnabled,
                                    darTemplatePublished,
                                    cohortEnabled,
                                },
                            })
                        }
                        size="large"
                        sx={{ p: 0 }}
                        iconSx={{ mr: 0 }}
                    />
                </div>
            );
        },
        header: () => null,
        size: 43,
    }),

    columnHelper.display({
        id: "name",
        cell: ({ row: { original } }) => {
            const { datasetId, name } = original;
            const linkHref = `/${RouteName.DATASET_ITEM}/${datasetId}`;

            return (
                <Link href={linkHref}>
                    <div style={{ textAlign: "center" }}>{name}</div>
                </Link>
            );
        },
        header: () => <span>{translations.name}</span>,
        size: 150,
    }),

    columnHelper.display({
        id: "darEnabled",
        cell: ({
            row: {
                original: { darEnabled },
            },
        }) => (
            <div style={{ textAlign: "center" }}>
                {darEnabled ? <CheckIcon color="primary" /> : "-"}
            </div>
        ),
        header: () => (
            <TooltipIcon
                buttonSx={{ p: 0 }}
                size="small"
                label={translations.darEnabled}
                content={translations.darEnabled}
            />
        ),
        size: 100,
    }),

    columnHelper.display({
        id: "cohortEnabled",
        cell: ({
            row: {
                original: { cohortEnabled },
            },
        }) => (
            <div style={{ textAlign: "center" }}>
                {cohortEnabled ? <CheckIcon color="primary" /> : "-"}
            </div>
        ),
        header: () => <span>{translations.cohortEnabled}</span>,
        size: 100,
    }),

    columnHelper.display({
        id: "dataCustodian",
        cell: ({
            row: {
                original: { dataCustodian },
            },
        }) => <div style={{ textAlign: "center" }}>{dataCustodian}</div>,
        header: () => <span>{translations.dataCustodian}</span>,
        size: 100,
    }),

    columnHelper.display({
        id: "entityType",
        cell: ({
            row: {
                original: { entityType },
            },
        }) => <div style={{ textAlign: "center" }}>{entityType}</div>,
        header: () => <span>{translations.entityType}</span>,
        size: 100,
    }),

    columnHelper.display({
        id: "delete",
        cell: ({ row: { original } }) => {
            return (
                <div style={{ textAlign: "center" }}>
                    <IconButton onClick={() => handleRemove(original.id)}>
                        <DeleteForeverIcon color="primary" />
                    </IconButton>
                </div>
            );
        },
        header: () => null,
        size: 43,
    }),
];

export { getColumns };
