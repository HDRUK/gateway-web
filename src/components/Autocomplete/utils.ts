import { ValueType } from "./Autocomplete";

const getChipLabel = (
    options: { value: string | number; label: string }[],
    value: ValueType
) => {
    if (typeof value === "string") {
        return value;
    }

    return options.find(option => option.value === value)?.label;
};

export { getChipLabel };
