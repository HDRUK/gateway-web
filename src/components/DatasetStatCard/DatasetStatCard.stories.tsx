import type { Meta, StoryObj } from "@storybook/react";
import DatasetStatCard from "./DatasetStatCard";

const meta: Meta<typeof DatasetStatCard> = {
    component: DatasetStatCard,
    title: "Cards/DatasetStatCard",
    tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof DatasetStatCard>;

export const Default: Story = {
    args: {
        title: "Population size",
        stat: "< 23",
        unit: "days",
        iconSrc: "/images/dataset/clock-white.svg",
    },
};
