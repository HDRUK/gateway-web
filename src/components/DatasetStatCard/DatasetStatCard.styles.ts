import { Card, styled } from "@mui/material";
import Image from "next/image";
import { colors } from "@/config/theme";

export const StatCard = styled(Card)(() => ({
    display: "flex",
    flexDirection: "column",
    background: `linear-gradient(90deg,${colors.purple400} calc(100% - 13px), ${colors.white} 13px)`,
    color: colors.white,
    borderRadius: 5,
    boxShadow: "3px 3px 6px 0px rgba(0,0,0,.09)",
    paddingRight: "13px",
    minHeight: "121px",
}));

export const Title = styled("div")(() => ({
    display: "flex",
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center",
}));

export const InfoWrapper = styled("div")(({ theme }) => ({
    display: "flex",
    flexDirection: "row",
    alignContent: "center",
    justifyContent: "center",
    padding: theme.spacing(1),
    paddingTop: 0,
    flex: 1,
}));

export const StatWrapper = styled("div")(() => ({
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
    justifyContent: "center",
    flex: "1 0 50%",
}));

export const StatImageWrapper = styled("div")(() => ({
    flex: "1 0 50%",
    position: "relative",
}));

export const StatImage = styled(Image)(({ theme }) => ({
    paddingLeft: theme.spacing(1),
}));
