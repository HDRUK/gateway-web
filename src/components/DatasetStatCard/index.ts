import DatasetInfoCard from "./DatasetStatCard";

export default DatasetInfoCard;
