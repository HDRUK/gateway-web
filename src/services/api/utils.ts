import { NotificationOptions } from "@/interfaces/Api";
import { AxiosResponse } from "axios";
import { Error } from "@/interfaces/Error";
import notificationService from "../notification";

interface ErrorNotificationProps {
    props: NotificationOptions;
    errorResponse: AxiosResponse<Error>;
    method: "delete" | "post" | "put" | "get";
}

const errorNotification = ({
    errorResponse,
    method,
    props,
}: ErrorNotificationProps) => {
    const { t, i18n, ...notificationProps } = props;
    const { data, status } = errorResponse || {};

    const fallbackTitle = i18n.exists(`api:common.error.status.${status}`)
        ? t(`api:common.error.status.${status}`)
        : "There has been an error";

    const title = data?.title || fallbackTitle;
    const message =
        data?.message ||
        t(`api:common.error.${method}.message`, {
            item: props.itemName || "Item",
        });

    notificationService.apiError(message, {
        title,
        message,
        errors: data?.errors,
        ...notificationProps,
    });
};

interface SuccessNotificationProps {
    props: NotificationOptions;
    method: "delete" | "post" | "put";
}

const successNotification = ({ props, method }: SuccessNotificationProps) => {
    const { t, i18n, ...notificationProps } = props;
    const customMessage = `api:${props.localeKey}.success.${method}.message`;
    const shouldOverideMessage = i18n.exists(customMessage);

    const message = shouldOverideMessage
        ? t(customMessage)
        : t(`api:common.success.${method}.message`, {
              item: props.itemName || "Item",
          });

    notificationService.success(message, { ...notificationProps });
};

export { errorNotification, successNotification };
