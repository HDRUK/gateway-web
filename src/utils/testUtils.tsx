import React, { ReactNode } from "react";
import {
    render,
    renderHook,
    RenderHookOptions,
    RenderOptions,
    RenderResult,
    RenderHookResult,
} from "@testing-library/react";
import { SWRConfig } from "swr";
import { CacheProvider, ThemeProvider } from "@emotion/react";
import ActionBarProvider from "@/providers/ActionBarProvider";
import DialogProvider from "@/providers/DialogProvider";
import theme from "@/config/theme";
import createEmotionCache from "@/config/createEmotionCache";

const clientSideEmotionCache = createEmotionCache();

const Wrapper = ({ children }: { children: ReactNode }) => {
    return (
        <SWRConfig
            value={{
                provider: () => new Map(),
            }}>
            <CacheProvider value={clientSideEmotionCache}>
                <ThemeProvider theme={theme}>
                    <ActionBarProvider>
                        <DialogProvider>{children}</DialogProvider>
                    </ActionBarProvider>
                </ThemeProvider>
            </CacheProvider>
        </SWRConfig>
    );
};

interface OptionProps
    extends RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRender = (
    ui: React.ReactElement<
        unknown,
        string | React.JSXElementConstructor<unknown>
    >,
    options?: OptionProps
): RenderResult => {
    const { wrapperProps, ...rest } = options || {};

    return render(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

interface OptionHookProps
    extends RenderHookOptions<
        unknown,
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRenderHook = (
    ui: (initialProps: unknown) => unknown,
    options?: OptionHookProps
): RenderHookResult<unknown, unknown> => {
    const { wrapperProps, ...rest } = options || {};
    return renderHook(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

export * from "@testing-library/react";

export { customRenderHook as renderHook };
export { customRender as render };
