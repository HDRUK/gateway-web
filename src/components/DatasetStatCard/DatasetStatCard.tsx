"use client";

import Typography from "@mui/material/Typography";
import { hasValidValue } from "@/utils/dataset";
import TooltipIcon from "../TooltipIcon";
import {
    InfoWrapper,
    StatCard,
    StatImage,
    StatImageWrapper,
    StatWrapper,
    Title,
} from "./DatasetStatCard.styles";

export interface DatasetStatCardProps {
    title: string;
    stat: string | string[];
    largeStatText?: boolean;
    iconSrc: string;
    unit?: string;
    helperText?: string;
    noStatText?: string;
}

const DatasetStatCard = ({
    title,
    stat,
    largeStatText,
    iconSrc,
    unit,
    helperText,
    noStatText,
}: DatasetStatCardProps) => {
    return (
        <StatCard>
            <Title>
                <Typography fontSize={16} sx={{ mb: 0, pt: 1, pb: 1 }}>
                    {title}
                </Typography>
                {helperText && (
                    <TooltipIcon content={helperText} label="" invertColor />
                )}
            </Title>

            <InfoWrapper>
                {hasValidValue(stat) ? (
                    <StatWrapper>
                        {Array.isArray(stat) ? (
                            stat.map((item, index) => (
                                <Typography
                                    fontSize={16}
                                    sx={{ alignSelf: "flex-start" }}
                                    key={`${stat}_${item}`}>
                                    {index < 2
                                        ? item
                                        : index === 3
                                        ? "..."
                                        : null}
                                </Typography>
                            ))
                        ) : (
                            <Typography fontSize={largeStatText ? 24 : 16}>
                                {stat}
                            </Typography>
                        )}
                        {hasValidValue(unit) && (
                            <Typography sx={{ pb: 1 }}>{unit}</Typography>
                        )}
                    </StatWrapper>
                ) : (
                    <StatWrapper>{noStatText}</StatWrapper>
                )}

                {iconSrc && (
                    <StatImageWrapper>
                        <StatImage src={iconSrc} alt="" fill />
                    </StatImageWrapper>
                )}
            </InfoWrapper>
        </StatCard>
    );
};

export default DatasetStatCard;
