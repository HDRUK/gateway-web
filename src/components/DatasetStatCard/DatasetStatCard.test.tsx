import { render } from "@/utils/testUtils";
import DatasetStatCard from "./DatasetStatCard";

describe("DatasetInfoCard", () => {
    it("should match snapshot", async () => {
        const img = "/test.jpg";
        const title = "A description";
        const stat = "10";
        const unit = "days";
        const helperText = "helper";

        const wrapper = render(
            <DatasetStatCard
                iconSrc={img}
                title={title}
                stat={stat}
                largeStatText
                unit={unit}
                helperText={helperText}
            />
        );

        expect(wrapper.container).toMatchSnapshot();
    });
});
