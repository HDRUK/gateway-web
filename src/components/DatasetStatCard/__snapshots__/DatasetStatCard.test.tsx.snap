// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetStatCard should match snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root mui-style-1tdwrob-MuiPaper-root-MuiCard-root"
  >
    <div
      class="mui-style-1hyoz7m"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 mui-style-1bwtdq5-MuiTypography-root"
      >
        A description
      </p>
      <div
        class="MuiBox-root mui-style-1lekzkb"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium mui-style-17wis5z-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-style-slk7cj-MuiSvgIcon-root"
            data-testid="InfoIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="mui-style-r7t1kn"
    >
      <div
        class="mui-style-1cfaonj"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 mui-style-qhyxbx-MuiTypography-root"
        >
          10
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 mui-style-qri3ce-MuiTypography-root"
        >
          days
        </p>
      </div>
      <div
        class="mui-style-n6cn66"
      >
        <img
          alt=""
          class="mui-style-1iuj5ih"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          sizes="100vw"
          src="/_next/image?url=%2Ftest.jpg&w=3840&q=75"
          srcset="/_next/image?url=%2Ftest.jpg&w=640&q=75 640w, /_next/image?url=%2Ftest.jpg&w=750&q=75 750w, /_next/image?url=%2Ftest.jpg&w=828&q=75 828w, /_next/image?url=%2Ftest.jpg&w=1080&q=75 1080w, /_next/image?url=%2Ftest.jpg&w=1200&q=75 1200w, /_next/image?url=%2Ftest.jpg&w=1920&q=75 1920w, /_next/image?url=%2Ftest.jpg&w=2048&q=75 2048w, /_next/image?url=%2Ftest.jpg&w=3840&q=75 3840w"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
      </div>
    </div>
  </div>
</div>
`;
